## Week 4 Assignment

In [2]:
#set up the basic package
import numpy as np

#### Task1

In [4]:
#Write a Python/NumPy code block that finds the distinct/unique common items between these two arrays

#follow instruction to set up two arrays.
a = np.array([1,2,3,2,3,4,3,4,5,6])
b= np.array([7,2,10,2,7,4,9,4,9,8])


# using np.intersect1d to find the distinct common values between two array
print(np.intersect1d(a,b))


[2 4]


Based on the result, distinct items are 2 and 4.

#### Task2

In [12]:
#Createthe following 5x3array using Python’s / NumPy’s  sequencing functionality 

#step 1: using one line of code to create array "a" ranged from 1 to 15,in 3*5 shape. 
a = np.arange(1,16).reshape(3,5)
# Reshape it by using transpose
print(a.T)





[[ 1  6 11]
 [ 2  7 12]
 [ 3  8 13]
 [ 4  9 14]
 [ 5 10 15]]


As the result shows, we get the array as the graph showed.

#### Task3

In [14]:
#Write a Python/Numpy code block that removes from array a any items that are also present in array b.
#follow instruction to set up two array first
a = np.array([12, 5, 7, 15, 3, 1, 8])
b = np.array([14, 6, 3, 11, 19, 12, 5])

#rephase the question, it equals asking the unique items in a.

# using np.setdiff1d to find unqiue array "uniq_a" which includes elements in a but not in b
uniq_a= np.setdiff1d(a,b)
# getthe result
uniq_a

array([ 1,  7,  8, 15])

results are array([ 1,  7,  8, 15])

#### Task4

Based on the research, the distance between New York City and Key West is 1547.4mile based on Google map.I choose 1547 as the distance value. Since 55 miles on day 1 and 65 miles on day 2 is given, other 33 days including 4 days of no travel need to be made up to fulfil the trip.Based on the inormation provided, 33 days are used for travel other 1427mile.
I tried random number generator for this problem but sometimes it provided non-sense value on a day since it difintely the limitation of bike travel per day(For example, travel more than 200 mile a day is not reasonable). Finally, I decided to make up the whole array.

In [18]:
#Given day 1 and day2 values, From day3, use random.randint method to generate 33 random numbers between 120 and 1547.

day3_end = np.random.randint(120, 1547,33)
day3_end
#make sure that there is four days without biking, which means the distance between two days keeps the same.
for i in np.random.randint(1,32,4):
    day3_end[i] = day3_end[i+1]
#make sure last day can finish the trip 
day3_end[32] = 1547
day3_end

#since it does not include first two days,now add firs two day value
beginning= np.array([55,120])
#use np.concatenate to add items 
total=np.concatenate((beginning, day3_end))
#reorder to make it in increasing order
travel = np.sort(total)
#get travel record
print(travel)


[  55  120  123  168  198  296  296  303  318  340  389  415  552  620  634
  655  665  703  704  704  758  763  763  787  905  924 1090 1090 1242 1271
 1372 1373 1526 1545 1547]


In [19]:
# using diff function to show the ride per day
travel_p_day = np.diff(travel)

# Show the total miles rode each day 
travel_p_day 

array([ 65,   3,  45,  30,  98,   0,   7,  15,  22,  49,  26, 137,  68,
        14,  21,  10,  38,   1,   0,  54,   5,   0,  24, 118,  19, 166,
         0, 152,  29, 101,   1, 153,  19,   2])

Therefore, this is the answer for this problem.

### DAV 5400 Week 4 Assignment (NumPy): Problem 5 Data Loader

In [17]:

#from template 
import pandas
import numpy as np

# supress scientific notation
np.set_printoptions(suppress = True)

# read the data file from GitHub and convert to a NumPy ndarray object
nyc_water = pandas.read_csv('https://raw.githubusercontent.com/jtopor/DAV-5400/master/Week4/Water_Consumption_In_The_New_York_City.csv').as_matrix()

# verify you have an ndarray object
type(nyc_water)

numpy.ndarray

#### Task5

In [18]:
#P1 find the maximum yearly NYC consumption of water in millions of gallons per day

# Create array and extract the water consumption from array "nyc_water"
water = nyc_water[:,2]

# using np.max to find the maximum value
np.max(water)

1512.0

The maximum yearly NYC consumption of water in millions of gallons per day is 1512 gallon.

In [20]:
#P2 Find how many calendar years are represented within this dataset


#using np.shape find calendar year
nyc_water.shape[0]

39

Based on the result, 39 calendar years are represented within this dataset.

In [24]:
#P3 Find the mean and the standard deviation of the per capita daily water consumption.

#create array for the per capita daily water consumption
water_per_cap = nyc_water[:,3]
#print mean and standard deviation of the array
print(np.mean(water_per_cap))
print(np.std(water_per_cap))

160.307692308
31.5189433037


the mean and the standard deviation of the per capita daily water consumption is 160.307692308 per capita daily water consumptio and 31.5189433037.

In [29]:
#P4 find the increase or decrease in population from year to year.
#population is on second column

# Use np.diff functions to find the difference on population.
pop_diff = np.diff(nyc_water[:,1], axis = 0)
print(pop_diff)

[ -30461.    17602.    19864.    72119.    53290.    39540.    45192.
   23230.    11243.    -9544.    -8525.    38851.    54443.    77222.
   64292.    62582.    64772.    75631.    84816.    89401.    60618.
   16685.5   16685.5   16685.5   16685.5   16685.5   16685.5   16685.5
   16685.5   16685.5   16685.5  108965.    77081.    61281.    49530.
   44512.    21171.    85025. ]


The result is show as screen